sketch/src/core/sketch_support.ts

```diff
@@ -0,0 +1,327 @@
+  return Math.floor(Math.random() * (b - a) + a);
+}
+export function randn(mu: number, std: number): number {
+  return mu + gaussRandom() * std;
+}
+// From http://www.math.grin.edu/~mooret/courses/math336/bivariate-normal.html
+export function birandn(
+  mu1: number,
+  mu2: number,
+  std1: number,
+  std2: number,
+  rho: number): number[] {
+  const z1 = randn(0, 1);
+  const z2 = randn(0, 1);
+  const x = Math.sqrt(1 - rho * rho) * std1 * z1 + rho * std1 * z2 + mu1;
+  const y = std2 * z2 + mu2;
+  return [x, y];
+}
+
+/**
+ * Sample from a categorial distribution
+ */
+export function sampleSoftmax(zSample: number[]): number {
+  const x = randf(0, 1);
+  const N = zSample.length;
+  let accumulate = 0;
+  let i: number;
+  for (i = 0; i < N; i++) {
+    accumulate += zSample[i];
+    if (accumulate >= x) {
+      return i;
+    }
+  }
+  console.log('error sampling pi index');
+  return -1;
+}
+
+// Data Tool Functions
+
+export function simplifyLine(V: number[][], tolerance=2.0): number[][] {
+  // from https://gist.github.com/adammiller/826148
+  // V ... [[x1,y1],[x2,y2],...] polyline
+  // tol  ... approximation tolerance
+  // ==============================================
+  // Copyright 2002, softSurfer (www.softsurfer.com)
+  // This code may be freely used and modified for any purpose
+  // providing that this copyright notice is included with it.
+  // SoftSurfer makes no warranty for this code, and cannot be held
+  // liable for any real or imagined damage resulting from its use.
+  // Users of this code must verify correctness for their application.
+  // http://softsurfer.com/Archive/algorithm_0205/algorithm_0205.htm
+
+  const tol = tolerance;
+
+  /*function sum(u: number[], v: number[]): number {
+    return [u[0] + v[0], u[1] + v[1]];
+  }*/
+  function diff(u: number[], v: number[]): number[] {
+    return [u[0] - v[0], u[1] - v[1]];
+  }
+  /*function prod(u: number[], v: number[]): number[] {
+    return [u[0] * v[0], u[1] * v[1]];
+  }*/
+  function dot(u: number[], v: number[]): number {
+    return u[0] * v[0] + u[1] * v[1];
+  }
+  function norm2(v: number[]): number {
+    return v[0] * v[0] + v[1] * v[1];
+  }
+  /*function norm(v: number[]): number {
+    return Math.sqrt(norm2(v));
+  }*/
+  function d2(u: number[], v: number[]): number {return norm2(diff(u, v));}
+  /*function d(u: number[], v: number[]): number {return norm(diff(u, v));}*/
+
+  function simplifyDP(tol: number, v: number[][],
+    j: number, k: number, mk: number[]) {
+    //  This is the Douglas-Peucker recursive simplification routine
+    //  It just marks vertices that are part of the simplified polyline
+    //  for approximating the polyline subchain v[j] to v[k].
+    //  mk[] ... array of markers matching vertex array v[]
+    if (k <= j + 1) { // there is nothing to simplify
+      return;
+    }
+    // check for adequate approximation by segment S from v[j] to v[k]
+    let maxi = j;          // index of vertex farthest from S
```

Comment:
Some of the formatting seems a bit off. Are you using clang-format?

Reply:
Okay let me look into this and fix.